In [2]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

In [3]:
# importiamo il primo file csv
artisti = pd.read_csv('Artists.csv', sep=',')

In [4]:
artisti.head()

Name                      ID Gender  Age Country  \
0        Drake  3TVXtAsR1Inumwj472S9r4   male   33      CA   
1  Post Malone  246dkjvS1zLTtiykXe5h60   male   25      US   
2   Ed Sheeran  6eUKZXaKkcviH0Ku9w2n3V   male   29      GB   
3     J Balvin  1vyhD5VmyZ7KMfW5gqLgo5   male   35      CO   
4    Bad Bunny  4q3ewBCX7sLwd24euuV69X   male   26      PR   

                                              Genres  Popularity  Followers  \
0  ['canadian hip hop', 'canadian pop', 'hip hop'...          95   83298497   
1           ['dfw rap', 'melodic rap', 'pop', 'rap']          86   43130108   
2         ['pop', 'singer-songwriter pop', 'uk pop']          87  115998928   
3  ['reggaeton', 'reggaeton colombiano', 'trap la...          83   38028010   
4      ['reggaeton', 'trap latino', 'urbano latino']          95   77931484   

                                     URI  
0  spotify:artist:3TVXtAsR1Inumwj472S9r4  
1  spotify:artist:246dkjvS1zLTtiykXe5h60  
2  spotify:artist:6eUKZXaKkcviH0Ku9w2n3V  
3  spotify:artist:1vyhD5VmyZ7KMfW5gqLgo5  
4  spotify:artist:4q3ewBCX7sLwd24euuV69X

In [4]:
artisti.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9488 entries, 0 to 9487
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Name        9488 non-null   object
 1   ID          9488 non-null   object
 2   Gender      7829 non-null   object
 3   Age         9488 non-null   int64 
 4   Country     6305 non-null   object
 5   Genres      9488 non-null   object
 6   Popularity  9488 non-null   int64 
 7   Followers   9488 non-null   int64 
 8   URI         9488 non-null   object
dtypes: int64(3), object(6)
memory usage: 667.3+ KB


In [4]:
# importiamo il secondo file csv
top_songs = pd.read_csv('Top_Songs_US.csv', sep=',')

In [6]:
top_songs.head(20)

Album Type               Artist ID  Artist Name  Artist Song Rank  \
0       album  3TVXtAsR1Inumwj472S9r4        Drake                 1   
1       album  3TVXtAsR1Inumwj472S9r4        Drake                 2   
2       album  3TVXtAsR1Inumwj472S9r4        Drake                 3   
3       album  3TVXtAsR1Inumwj472S9r4        Drake                 4   
4       album  3TVXtAsR1Inumwj472S9r4        Drake                 5   
5       album  3TVXtAsR1Inumwj472S9r4        Drake                 6   
6       album  3TVXtAsR1Inumwj472S9r4        Drake                 7   
7       album  3TVXtAsR1Inumwj472S9r4        Drake                 8   
8       album  3TVXtAsR1Inumwj472S9r4        Drake                 9   
9       album  3TVXtAsR1Inumwj472S9r4        Drake                10   
10      album  246dkjvS1zLTtiykXe5h60  Post Malone                 1   
11     single  246dkjvS1zLTtiykXe5h60  Post Malone                 2   
12      album  246dkjvS1zLTtiykXe5h60  Post Malone                 3   
13      album  246dkjvS1zLTtiykXe5h60  Post Malone                 4   
14     single  246dkjvS1zLTtiykXe5h60  Post Malone                 5   
15      album  246dkjvS1zLTtiykXe5h60  Post Malone                 6   
16      album  246dkjvS1zLTtiykXe5h60  Post Malone                 7   
17      album  246dkjvS1zLTtiykXe5h60  Post Malone                 8   
18      album  246dkjvS1zLTtiykXe5h60  Post Malone                 9   
19      album  246dkjvS1zLTtiykXe5h60  Post Malone                10   

                                       Track Name  Is Playable  \
0                              IDGAF (feat. Yeat)         True   
1            First Person Shooter (feat. J. Cole)         True   
2         Rich Baby Daddy (feat. Sexyy Red & SZA)         True   
3                   Jimmy Cooks (feat. 21 Savage)         True   
4                                       One Dance         True   
5                          MELTDOWN (feat. Drake)         True   
6                                  Virginia Beach         True   
7                                       Rich Flex         True   
8                       Slime You Out (feat. SZA)         True   
9                                      God's Plan         True   
10  Sunflower - Spider-Man: Into the Spider-Verse         True   
11                                       Chemical         True   
12                                        Circles         True   
13    I Like You (A Happier Song) (with Doja Cat)         True   
14                  Dial Drunk (with Post Malone)         True   
15                     rockstar (feat. 21 Savage)         True   
16                                Congratulations         True   
17                                           Wow.         True   
18                                     Better Now         True   
19                                   I Fall Apart         True   

                       Album Name Release Date  Total Album Tracks  \
0                For All The Dogs   2023-10-06                  23   
1                For All The Dogs   2023-10-06                  23   
2                For All The Dogs   2023-10-06                  23   
3             Honestly, Nevermind   2022-06-17                  14   
4                           Views   2016-05-06                  20   
5                          UTOPIA   2023-07-28                  19   
6                For All The Dogs   2023-10-06                  23   
7                        Her Loss   2022-11-04                  16   
8                For All The Dogs   2023-10-06                  23   
9                        Scorpion   2018-06-29                  25   
10           Hollywood's Bleeding   2019-09-06                  17   
11                       Chemical   2023-04-14                   1   
12           Hollywood's Bleeding   2019-09-06                  17   
13         Twelve Carat Toothache   2022-06-03                  14   
14  Dial Drunk (with Post Malone)   2023-07-18       

In [10]:
top_songs.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 37146 entries, 0 to 37145
Data columns (total 16 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   Album Type          37146 non-null  object
 1   Artist ID           37146 non-null  object
 2   Artist Name         37146 non-null  object
 3   Artist Song Rank    37146 non-null  int64 
 4   Track Name          37146 non-null  object
 5   Is Playable         37146 non-null  bool  
 6   Album Name          37146 non-null  object
 7   Release Date        37146 non-null  object
 8   Total Album Tracks  37146 non-null  int64 
 9   Is Explicit         37146 non-null  bool  
 10  ISRC                37146 non-null  object
 11  Song Duration       37146 non-null  int64 
 12  Track Number        37146 non-null  int64 
 13  Popularity          37146 non-null  int64 
 14  Track Id            37146 non-null  object
 15  Track URI           37146 non-null  object
dtypes: bool(2), int64(5), 

In [5]:
# convertiamo la colonna 'Release Date' in datetime
top_songs['Release Date'] = pd.to_datetime(top_songs['Release Date'], utc=True).dt.tz_convert('Europe/Berlin')

In [8]:
top_songs.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 37146 entries, 0 to 37145
Data columns (total 16 columns):
 #   Column              Non-Null Count  Dtype                        
---  ------              --------------  -----                        
 0   Album Type          37146 non-null  object                       
 1   Artist ID           37146 non-null  object                       
 2   Artist Name         37146 non-null  object                       
 3   Artist Song Rank    37146 non-null  int64                        
 4   Track Name          37146 non-null  object                       
 5   Is Playable         37146 non-null  bool                         
 6   Album Name          37146 non-null  object                       
 7   Release Date        37146 non-null  datetime64[ns, Europe/Berlin]
 8   Total Album Tracks  37146 non-null  int64                        
 9   Is Explicit         37146 non-null  bool                         
 10  ISRC                37146 non-null

In [39]:
top_songs.head(5) # osserviamo come è cambiata la colonna 'Release Date'

Album Type               Artist ID Artist Name  Artist Song Rank  \
0      album  3TVXtAsR1Inumwj472S9r4       Drake                 1   
1      album  3TVXtAsR1Inumwj472S9r4       Drake                 2   
2      album  3TVXtAsR1Inumwj472S9r4       Drake                 3   
3      album  3TVXtAsR1Inumwj472S9r4       Drake                 4   
4      album  3TVXtAsR1Inumwj472S9r4       Drake                 5   

                                Track Name  Is Playable           Album Name  \
0                       IDGAF (feat. Yeat)         True     For All The Dogs   
1     First Person Shooter (feat. J. Cole)         True     For All The Dogs   
2  Rich Baby Daddy (feat. Sexyy Red & SZA)         True     For All The Dogs   
3            Jimmy Cooks (feat. 21 Savage)         True  Honestly, Nevermind   
4                                One Dance         True                Views   

  Release Date  Total Album Tracks  Is Explicit          ISRC  Song Duration  \
0   2023-10-06                  23         True  USUG12306072         260111   
1   2023-10-06                  23         True  USUG12306071         247444   
2   2023-10-06                  23         True  USUG12306085         319191   
3   2022-06-17                  14         True  USUG12204897         218364   
4   2016-05-06                  20        False  USCM51600028         173986   

   Track Number  Popularity                Track Id  \
0             7          93  2YSzYUF3jWqb9YP9VXmpjE   
1             6          91  7aqfrAY2p9BUSiupwk3svU   
2            20          89  1yeB8MUNeLo9Ek1UEpsyz6   
3            14          89  3F5CgOj3wFlRv51JsHbxhe   
4            12          89  1zi7xx7UVEFkmKfv06H8x0   

                              Track URI  
0  spotify:track:2YSzYUF3jWqb9YP9VXmpjE  
1  spotify:track:7aqfrAY2p9BUSiupwk3svU  
2  spotify:track:1yeB8MUNeLo9Ek1UEpsyz6  
3  spotify:track:3F5CgOj3wFlRv51JsHbxhe  
4  spotify:track:1zi7xx7UVEFkmKfv06H8x0

Analizziamo

Quale è il genere più ascoltato?

In [6]:
# Facciamo il merge dei due file tramite l'ID dell'artista
merged_df = pd.merge(artisti, top_songs, left_on='ID', right_on='Artist ID')

In [7]:
# popolarità dell'artista
artist_popularity = merged_df.groupby('Genres')['Popularity_x'].sum() # raggruppiamo per genere e sommiamo la popolarità per ciascun genere

In [8]:
# popolarità delle canzoni
song_popularity = merged_df.groupby('Genres')['Popularity_y'].sum() # raggruppiamo per genere e sommiamo la popolarità per ciascun genere

In [9]:
# visualizziamo il genere dell'artista più ascoltato/popolare
most_listened_artist_genre = artist_popularity.idxmax()
most_listened_artist_genre

"['pop']"

In [10]:
# visualizziamo il genere delle canzoni più ascoltate/popolari
most_listened_song_genre = song_popularity.idxmax()
most_listened_song_genre

"['pop']"

Quale è la canzone più ascoltata? (che ha più popolarità)

In [11]:
most_listened_song = top_songs.loc[top_songs['Popularity'].idxmax()]

most_listened_song[['Track Name', 'Artist Name', 'Popularity']]

Track Name     Cruel Summer
Artist Name    Taylor Swift
Popularity               99
Name: 550, dtype: object

Quale è invece la canzone meno ascoltata? (che ha meno popolarità)

In [12]:
least_listened_song = top_songs.loc[top_songs['Popularity'].idxmin()]

least_listened_song[['Track Name', 'Artist Name', 'Popularity']]

Track Name     Ride Or Die
Artist Name           Nyla
Popularity               0
Name: 3107, dtype: object

Analizziamo la correlazione tra la popolarità di un artista e il gender (il sesso) dell'artista. Osserviamo l'artista più famoso

In [13]:
popolarità_artista = merged_df.groupby(['ID', 'Name', 'Gender'])['Popularity_x'].sum()

most_popular_artist = popolarità_artista.idxmax()
most_popular_artist

('06HL4z0CvFAxyc27GXpf02', 'Taylor Swift', 'female')

In [14]:
# ora osserviamo la top 5 artisti più famosi
top_5_artists = popolarità_artista.nlargest(5)
top_5_artists

ID                      Name          Gender
06HL4z0CvFAxyc27GXpf02  Taylor Swift  female    10000
3TVXtAsR1Inumwj472S9r4  Drake         male       9500
4q3ewBCX7sLwd24euuV69X  Bad Bunny     male       9500
1Xyo4u8uXC1ZmMpatF05PJ  The Weeknd    male       9300
0Y5tJX1MQlPlqiwlOH1tJY  Travis Scott  male       9000
Name: Popularity_x, dtype: int64

In [15]:
# ora una top 10 artisti più famosi
top_10_artists = popolarità_artista.nlargest(10)
top_10_artists

ID                      Name           Gender
06HL4z0CvFAxyc27GXpf02  Taylor Swift   female    10000
3TVXtAsR1Inumwj472S9r4  Drake          male       9500
4q3ewBCX7sLwd24euuV69X  Bad Bunny      male       9500
1Xyo4u8uXC1ZmMpatF05PJ  The Weeknd     male       9300
0Y5tJX1MQlPlqiwlOH1tJY  Travis Scott   male       9000
5K4W6rqBFWDnAN6FQUkS6x  Kanye West     male       9000
7dGJo4pcD2V6oG8kP0tJRR  Eminem         male       8900
1RyvyyTE3xzB2ZywiAwp0i  Future         male       8800
1URnnhqYAYcrqrcwql10ft  21 Savage      male       8800
1uNFoZAHBGtllmzznpCI3s  Justin Bieber  male       8800
Name: Popularity_x, dtype: int64

Da questa analisi, osserviamo che l'artista più ascoltato è Taylor Swift, che ha come genere "female" (femminile). Per il resto della classifica, gli artisti presenti hanno come genere "male" (maschile). Dunque, possiamo dedurre che la maggior parte degli artisti più ascoltati è di genere maschile.

Adesso, come sotto analisi, osserviamo il genere degli artisti di genere maschile.

In [16]:
popolarità_artista = merged_df.groupby(['ID', 'Name', 'Gender', 'Genres'])['Popularity_x'].sum()

top_10_with_genre = popolarità_artista.nlargest(10)
top_10_with_genre

ID                      Name           Gender  Genres                                                           
06HL4z0CvFAxyc27GXpf02  Taylor Swift   female  ['pop']                                                              10000
3TVXtAsR1Inumwj472S9r4  Drake          male    ['canadian hip hop', 'canadian pop', 'hip hop', 'pop rap', 'rap']     9500
4q3ewBCX7sLwd24euuV69X  Bad Bunny      male    ['reggaeton', 'trap latino', 'urbano latino']                         9500
1Xyo4u8uXC1ZmMpatF05PJ  The Weeknd     male    ['canadian contemporary r&b', 'canadian pop', 'pop']                  9300
0Y5tJX1MQlPlqiwlOH1tJY  Travis Scott   male    ['hip hop', 'rap', 'slap house']                                      9000
5K4W6rqBFWDnAN6FQUkS6x  Kanye West     male    ['chicago rap', 'hip hop', 'rap']                                     9000
7dGJo4pcD2V6oG8kP0tJRR  Eminem         male    ['detroit hip hop', 'hip hop', 'rap']                                 8900
1RyvyyTE3xzB2ZywiAwp0i  Future   

Da questa analisi vediamo che, oltre al pop, i generi più ascoltati sono 'hip hop' e 'rap', insieme ai loro sotto generi (e.g 'atl hip hop' e 'pop rap')

In [17]:
sorted_songs = top_songs.sort_values(by='Popularity', ascending=False)
top_10_songs = sorted_songs.drop_duplicates(subset='Track Name').head(15)
top_10_songs[['Track Name', 'Popularity', 'Is Explicit']]

Track Name  Popularity  \
36856                                       Cruel Summer          99   
3018                     All I Want for Christmas Is You          99   
29988                                        PERRO NEGRO          97   
41                                                MONACO          96   
28308                                      Santa Tell Me          95   
28598                                       HARLEY QUINN          95   
25394                                               LALA          95   
20300  What Was I Made For? [From The Motion Picture ...          94   
20281                                            Starboy          94   
20280     One Of The Girls (with JENNIE, Lily Rose Depp)          94   
8619   Is It Over Now? (Taylor's Version) (From The V...          94   
5104                                           As It Was          94   
17316                                           fukumean          93   
35330                                              Prada          93   
32462                                            Houdini          93   

       Is Explicit  
36856        False  
3018         False  
29988         True  
41            True  
28308        False  
28598         True  
25394         True  
20300        False  
20281         True  
20280        False  
8619         False  
5104         False  
17316         True  
35330         True  
32462        False

In [18]:
conteggio = top_10_songs['Is Explicit'].value_counts()
conteggio

False    8
True     7
Name: Is Explicit, dtype: int64

In [19]:
# contiamo il totale di canzoni non esplicite (valore 'False') e il totale di canzoni esplicite (valore 'True')
explicit_songs = top_songs['Is Explicit'].value_counts()
explicit_songs

False    23196
True     13950
Name: Is Explicit, dtype: int64

Su un totale di 23.196 canzoni non esplicite e un totale di 13.950 di canzoni esplicite, è chiaro che le canzoni non esplicite sono presenti in maggior numero. Inoltre, se osserviamo l'analisi precedente, su un totale di 15 canzoni con popolarità maggiore, la maggior parte di esse non sono esplicite. Infatti, contiamo 8 non esplicite e 7 esplicite. Dunque, possiamo concludere che gli ascoltatori preferiscono evidentemente ascoltare canzoni non esplicite.

Vediamo se è presente una correlazione tra l'età degli artisti e il numero di followers che hanno

In [20]:
correlation = artisti['Age'].corr(artisti['Followers'])
correlation

0.10171870524373149

Il coefficiente di correlazione ottenuto è approssimativamente 0.10, ciò significa che è presente una correlazione molto debole e positiva. Infatti, se è presente un trend positivo, la relazione non è molto forte. Possiamo quindi dedurre che l'età non costituisca un fattore che influisce molto sul numero di followers di un artista. 

Proviamo con un altro fattore

In [21]:
correlation_2 = artisti['Country'].astype('category').cat.codes.corr(artisti['Followers'])
correlation_2

0.11406598755198748

Anche qui, come la precedente analisi, possiamo dedurre che la nazione di provenienza di un artista non sia il fattore più determinante per quanto riguarda il numero di followers. Essendo però leggermente più alto, può significare che rispetto all'età, la nazione di provenienza possa influire di più sul numero di followers, ma resta comunque una relazione non molto forte.

Vediamo un ultimo esempio

In [22]:
artisti['num_genres'] = artisti['Genres'].str.split(',').apply(lambda x: len(x))
correlation_3 = artisti['num_genres'].corr(artisti['Followers'])
correlation_3

0.1913351681627288

In questo caso, come i precedenti, notiamo che la relazione tra le due variabili non è molto forte, ma in questo esempio il coefficiente di correlazione è più alto (approx. 0.19). Ciò può significare che gli artisti con più generi tendono ad avere più followers. 
È interessante notare come la varietà di generi di un artista possa avere un determinato ruolo nella popolarità di un artista, su dunque il numero dei followers. 